quinn/src/lib.rs

```diff
@@ -330,50 +248,41 @@ impl EndpointInner {
         incoming: mpsc::Sender<NewConnection>,
         ipv6: bool,
     ) -> Self {
+        let (sender, events) = mpsc::unbounded();
         Self {
             log,
             socket,
             inner,
             incoming,
             ipv6,
-            outgoing: None,
-            pending: FnvHashMap::default(),
-            timers: FuturesUnordered::new(),
+            sender,
+            events,
+            outgoing: VecDeque::new(),
             buffered_incoming: VecDeque::new(),
             driver: None,
+            connections: FnvHashMap::default(),
         }
     }
 
-    /// Wake up a blocked `Driver` task to process I/O
-    fn notify(&self) {
-        if let Some(x) = self.driver.as_ref() {
-            x.notify();
-        }
-    }
-
-    fn drive_recv(
-        &mut self,
-        now: Instant,
-        wrapped: Rc<RefCell<EndpointInner>>,
-    ) -> Result<bool, io::Error> {
+    fn drive_recv(&mut self, now: Instant) -> Result<bool, io::Error> {
         let mut buf = [0; 64 * 1024];
         let mut recvd = 0;
         loop {
             match self.socket.poll_recv(&mut buf) {
                 Ok(Async::Ready((n, addr, ecn))) => {
-                    if let Some(ch) = self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
-                        self.pending.insert(ch, Pending::new(None));
-                        match self.incoming.poll_ready() {
-                            Ok(Async::Ready(())) => {
-                                self.incoming
-                                    .start_send(NewConnection::new(wrapped.clone(), ch))
-                                    .unwrap();
-                                self.inner.accept();
-                            }
-                            _ => {
-                                self.buffered_incoming.push_back(ch);
-                            }
+                    match self.inner.handle(now, addr, ecn, (&buf[0..n]).into()) {
+                        Some((handle, EndpointAppEvent::NewConnection(conn))) => {
+                            let conn = self.create_connection(handle, conn);
+                            self.buffered_incoming.push_back((handle, conn));
```

Comment:
It's not ideal that a busy application will let connections languish in the incoming queue (either in `buffered_incoming` or in the `incoming` channel itself) where they won't be driven and may time out. This causes peers to waste time on spurious `Initial` retransmits, and may cause spurious timeouts.

We could solve this by doing away with buffering entirely, which isn't ideal for performance and is impossible with our current channel implementation anyway, or by writing a custom channel that tracks which connections have not yet been yielded to the application, allowing us to automatically drive them, perhaps from a `VecDeque` owned by the endpoint driver.

Reply:
With `drive_incoming()` running directly after `drive_recv()` and the latter being constrained by `IO_LOOP_BOUND`, can we agree that this can wait until after this PR?

Reply:
The problem arises if the application isn't reading from `incoming` at all, so a connection can remain there indefinitely. That said, it's probably okay to defer since it's a subtle inefficiency that only affects already-overloaded (or severely buggy) systems.